derive.h

```diff
@@ -0,0 +1,291 @@
+
+class BinaryExpression;
+shared_ptr<Expression> make_shared_expression(BinaryExpression* ptr);
+shared_ptr<Expression> make_shared_expression(CoefExpression* ptr);
+shared_ptr<Expression> multiply(shared_ptr<Expression> lhs, shared_ptr<Expression> rhs);
+shared_ptr<Expression> add(shared_ptr<Expression> lhs, shared_ptr<Expression> rhs);
+shared_ptr<Expression> pow(shared_ptr<Expression> lhs, shared_ptr<Expression> rhs);
+shared_ptr<Expression> subtract(shared_ptr<Expression> lhs, shared_ptr<Expression> rhs);
+
+class BinaryExpression : public Expression {
+public:
+	virtual ostream& printTo(ostream& os) const {
+		os << "(";
+		lhs->printTo(os);
+	 	os << operand;
+		rhs->printTo(os);
+		os << ")";
+		return os;
+	}
+
+	virtual bool is_zero() {
+		switch (operand) {
+			case plus: 
+				return lhs->is_zero() && rhs->is_zero();
+			case minus: 
+				return lhs->is_zero() && rhs->is_zero();
+			case mult: 
+				return lhs->is_zero() || rhs->is_zero();
+			case power:
+				return lhs->is_zero();
+		}
+		assert(0);
+		return false;
+	}
+
+	
+	BinaryExpression(shared_ptr<Expression> a_lhs, shared_ptr<Expression> a_rhs, Operand a_operand):
+		lhs(a_lhs),
+		rhs(a_rhs),
+		operand(a_operand) 
+	{ }
+	
+	virtual shared_ptr<Expression> derivative() const {
+		switch (operand) {
+			case plus: 
+				return add(lhs->derivative(), rhs->derivative());
+			case minus: 
+				return subtract(lhs->derivative(), rhs->derivative());
+			case mult: 
+				return add(
+					multiply(lhs->derivative(), rhs),
+					multiply(rhs->derivative(), lhs)
+				);
+			case power: 
+				{
+					shared_ptr<CoefExpression> coefExp = std::dynamic_pointer_cast<CoefExpression>(rhs);
+					assert(coefExp);
+					auto be1 = multiply(
+						rhs,
+						pow(
+							lhs, 
+							make_shared_expression(
+								new CoefExpression(coefExp->coef - 1)
+							)
+						)
+					);
+					return multiply(be1, lhs->derivative());
+				}
+			default: assert(0); 
+		}
+	}
+	shared_ptr<Expression> lhs;
+	shared_ptr<Expression> rhs;
+	Operand operand;
+};
+
+shared_ptr<Expression> make_shared_expression(CoefExpression* ptr) {
+	return std::dynamic_pointer_cast<Expression>(
+		shared_ptr<CoefExpression>(ptr)
+	);
+}
+
+shared_ptr<Expression> make_shared_expression(BinaryExpression* ptr) {
+	return std::dynamic_pointer_cast<Expression>(
+		shared_ptr<BinaryExpression>(ptr)
+	);
+}
+
+shared_ptr<Expression> add(shared_ptr<Expression> lhs, shared_ptr<Expression> rhs) {
+	auto coef_left = std::dynamic_pointer_cast<CoefExpression>(lhs);
+	auto coef_right = std::dynamic_pointer_cast<CoefExpression>(rhs);
+	if (coef_left && coef_right) {
+		return make_shared_expression(
+			new CoefExpression{coef_left->coef + coef_right->coef}
+		); 
+	}
+
+	if (lhs->is_zero()) {
+		return rhs;
+	}
+	
+	if (rhs->is_zero()) {
+		return lhs;
+	}
+	
+	return make_shared_expression(
+		new BinaryExpression{lhs, rhs, plus}
+	);
+}
+
+shared_ptr<Expression> subtract(shared_ptr<Expression> lhs, shared_ptr<Expression> rhs) {
```

Comment:
Looks like almost copy-paste from `add`, doesn't it?

Reply:
Of course. I thought to convert the -x to +(-1)*x, but I am not sure it worth it.